url0307c0e995_js_labs_js_collider_tgz-pJ8-org_jsl_tests_shmem_throughput_MainJ8/src/org/jsl/collider/SessionImpl.java

```diff
@@ -0,0 +1,1130 @@
+                    assert( m_iov[m_iovc] == null );
+                    m_iov[m_iovc] = node.buf.duplicate();
+                    m_iovc++;
+                    node = node.next;
+                }
+            }
+            else
+                joinMessages();
+
+            try
+            {
+                final long bytesSent = m_socketChannel.write( m_iov, 0, m_iovc );
+                if (bytesSent == 0)
+                {
+                    m_collider.executeInSelectorThread( m_starter );
+                    return;
+                }
+            }
+            catch (final IOException ex)
+            {
+                closeAndCleanupQueue( ex );
+                releaseSocket( "SocketWriter");
+                return;
+            }
+            catch (final NotYetConnectedException ex)
+            {
+                closeAndCleanupQueue( ex );
+                releaseSocket( "SocketWriter" );
+                return;
+            }
+
+            Node node = m_head;
+            for (int idx=0;;)
+            {
+                if (m_iov[idx].remaining() > 0)
+                {
+                    final int iovc = (m_iovc - idx);
+                    int cc = 0;
+                    for (; idx<m_iovc; idx++, cc++)
+                        m_iov[cc] = m_iov[idx];
+                    for (; cc<m_iovc; cc++)
+                        m_iov[cc] = null;
+                    m_iovc = iovc;
+                    m_head = node;
+                    m_collider.executeInThreadPool( this );
+                    return;
+                }
+
+                node.buf = null;
+                if (node.rbuf != null)
+                {
+                    node.rbuf.release();
+                    node.rbuf = null;
+                }
+
+                m_iov[idx] = null;
+                if  (++idx == m_iovc)
+                    break;
+
+                final Node next = node.next;
+                s_nodeNextUpdater.lazySet( node, null );
+                node = next;
+            }
+
+            m_iovc = 0;
+            final Node next = node.next;
+            if (next == null)
+            {
+                m_head = null;
+                if (!m_tail.compareAndSet(node, null))
+                {
+                    while (node.next == null);
+                    m_head = node.next;
+                    s_nodeNextUpdater.lazySet( node, null );
+                    if (m_head == CLOSE_MARKER)
+                        releaseSocket( "SocketWriter.runInThreadPool()" );
+                    else
+                        m_collider.executeInThreadPool( m_writer );
+                }
+            }
+            else
+            {
+                s_nodeNextUpdater.lazySet( node, null );
+                m_head = next;
+                if (m_head == CLOSE_MARKER)
+                    releaseSocket( "SocketWriter.runInThreadPool()" );
+                else
+                    m_collider.executeInThreadPool( m_writer );
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 327:349, 1106:1128.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.